### PR TITLE
Speed up logging of multiple metrics

### PR DESCRIPTION
**Pull request by creinders**
## What changes are proposed in this pull request?

Logging multiple metrics via log_batch takes very long when using SQLAlchemy storage (as described here #3010). 

With this pr, metrics are added batch-wise. In this way, only one session is opened which saved a lot of time. Additionally, storing the latest metrics is performed once at the end, so that only one database write is necessary.

It should be noted that metrics with the same run-id, step, timestamp, and value should be added only once. I assume that this conflict happens not that often. The check has originally been performed when adding the metric. However, it has turned out that this can be speed up by adding all metrics and catching possible integrity errors. If an integrity error appears, the metrics will be added in the slower mode (on after another).

Inserting 1000 new metrics now takes about 0.2 seconds instead of 9 seconds. If some metrics are duplicated, it takes about 2 seconds in the slower mode.

## How is this patch tested?

Covered by unit tests.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by creinders** [comment]
@ankitmathur-db as proposed here https://github.com/mlflow/mlflow/issues/3010. What do you think?

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [inline mlflow/store/tracking/sqlalchemy_store.py:496]: why was this special logic required?

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [inline mlflow/store/tracking/sqlalchemy_store.py:505]: these conditions sound contradictory. They are prolly required due to shared session. Please add a comment to clarify that

Comment (mparkhe) [reply mlflow/store/tracking/sqlalchemy_store.py:505]: Also please add a comment on why _all_ metrics are being grouped by the metric name -- a bit hard for reader to follow until they see call to `_update_latest_metric_if_necessary`

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [inline mlflow/store/tracking/sqlalchemy_store.py:474]: **nit** can you remove these additional blank lines -- makes the function unnecessarily long.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [inline mlflow/store/tracking/sqlalchemy_store.py:518]: please put a comment/ToDo that this part is currently not done in batch due to complexity in `_update_latest_metric_if_necessary` step

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [inline mlflow/store/tracking/sqlalchemy_store.py:549]: `None` metrics should be an error case and should have been handled in the caller function

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [inline mlflow/store/tracking/sqlalchemy_store.py:490]: I am not sure I understand this comment. Also unclear why `auto_new` is required

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [inline mlflow/store/tracking/sqlalchemy_store.py:732]: this change only batches metrics -- params and tags are still added one at a time.

I am curious -- was there a particular reason you chose not to implement batch insert for params and tags. Would be easier than metrics.

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (mparkhe) [inline tests/store/tracking/test_sqlalchemy_store.py:1477]: we should also add some additional tests to make sure that batched metric insertion is done correctly. For instance multiple metrics with the same key in a single batch, metrics with non-increasing steps, metrics from the batch **do not overwrite** existing metrics (with different keys or same keys if the existing ones have higher step values or timestamp) and **do overwrite** (existing metrics with same keys and lower timestamps and step values), ... etc.

**Review by mparkhe — CHANGES_REQUESTED** [review]
I added a few comments inlined. Would help to add some new recommended tests to verify functionality has not changed.

Also not sure why `auto_new` is required

**Comment by mparkhe** [comment]
Hi @creinders:  I have added some comments and requested changes to your pull request. Would you be able to address this in the coming week? This would a good change to be included in the next release.

**Comment by creinders** [comment]
> Hi @creinders: I have added some comments and requested changes to your pull request. Would you be able to address this in the coming week? This would a good change to be included in the next release.

Thanks for the comments, I will push an update the next days

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py]: The name of the variable was not very clear. I renamed auto_new to batch_mode. In batch mode, all metrics are added to the session and the constraints are checked during saving. This is much faster.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py:574]: I added a comment in the code for that. Regarding your first comment, I not sure I understand it correctly. In line 504, it will be checked if the metric is new (this might be not the case, if batch mode is disabled. In this case the metric with same step, value, ts, and name is already stored). Line 505: the metrics are store in a dictionary with the name as the key. If the key does not exist, an empty list will be added to the dictionary to collect all metrics.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py]: Done :)

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py]: I added a todo

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py]: The method will now check if an error should be raised

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py]: I improved the comments and added some details.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py:831]: Yes definitely - this can be done in the same way for params and tags as well. I just started with metrics because it influences the runtime the most. Handling the constraints might be a little bit more tricky for params and tags.

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (creinders) [reply tests/store/tracking/test_sqlalchemy_store.py:1806]: I added a couple of tests.

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (mparkhe) [inline tests/store/tracking/test_sqlalchemy_store.py:685]: also assert that timestamp and step values are correct

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (mparkhe) [inline tests/store/tracking/test_sqlalchemy_store.py:687]: i don't see the deactivation logic. looks like this test is exactly the same as above

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (mparkhe) [inline tests/store/tracking/test_sqlalchemy_store.py:815]: same comment as before. i am not sure how this is deactivating batch_mode

**Review by mparkhe — COMMENTED** [review]
The tests look good. The only thing that I am confused about is why support both `batch_mode` and non-`batch_mode`. Can the code be simplified. Please find more comments around `log_metrics` method.

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (mparkhe) [reply mlflow/store/tracking/sqlalchemy_store.py:545]: oh! i see ... this is `sqlalchemy.exc.IntegrityError` section

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py:571]: Unfortunately, `_get_or_create` is much slower because an additional query operation is performed for every metric (checking if it already exists). The same check is performed however when saving a session (due to the constraint check).

Using only `_get_or_create` takes about 2 seconds for 2000 points while inserting in this way takes only 0.2 seconds (if no conflicts are in the data which I guess is the case most of the time).

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (creinders) [reply mlflow/store/tracking/sqlalchemy_store.py:629]: Actually, the error will never arise due to previous checks in log_metric(s). `_update_latest_metric_if_necessary` is only called when it has at least one metric so that this case never occurs. Should when then switch back to just returning? Or do you have another recommendation?

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (creinders) [reply tests/store/tracking/test_sqlalchemy_store.py:685]: I added the checks

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (creinders) [reply tests/store/tracking/test_sqlalchemy_store.py:692]: Batch mode is deactivate because there are two metrics with same value, ts, step. In this way, the integrity check is tested.
In this test, the comment was missing which I added now to the definition of the metrics.

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (creinders) [reply tests/store/tracking/test_sqlalchemy_store.py:825]: There is a comment in line 848. Please also see the previous comment.

**Comment by creinders** [comment]
Hi @mparkhe, I addressed all points a while ago. What do you think?

**Comment by nickvazz** [comment]
Hi, I was wondering if this is still being worked on. Thank you!

**Comment by creinders** [comment]
Actually I was waiting for a response. The feature was fully implemented.

**Comment by mparkhe** [comment]
I will look at this again shortly.

@creinders : Thank you for your patience. Do you mind if i make some small edits to the pull request (if any required) before merging?

**Comment by creinders** [comment]
> @creinders : Thank you for your patience. Do you mind if i make some small edits to the pull request (if any required) before merging?
Sure you can make changes to the pull request

**Comment by lostless13** [comment]
Any chance of progressing this? We're really struggling with performance in this area.

**Comment by prakhar19** [comment]
Any updates on this? It is a much-needed feature. MLFlow is almost unusable for large batch logging.

**Comment by WeichenXu123** [comment]
@prakhar19 
I merge this PR code with master branch.
You can install the mflow with this PR fix by:
```
pip install git+https://github.com/creinders/mlflow.git@master
```

I will take over this PR to complete it.

**Comment by jaymegordo** [comment]
Does this PR only handle batch logging `metrics` but not `params`? Shouldn't they be handled the same way?

**Comment by WeichenXu123** [comment]
@jaymegordo
Yes. We should both handle metric and params. I will do that. Though metrics are more likely to be the perf bottleneck.

**Comment by erensahin** [comment]
Hi, I was thinking about adding this support and found this one. I can take over this PR and continue the work on it if no-one interests.

I also think anything should be logged in batches, not only metrics and params, also tags etc.

**Comment by harupy** [comment]
Addressed by https://github.com/mlflow/mlflow/pull/5460

**Pull request closed without merging**
